### PR TITLE
Merge Shopper Insights Feature Into Main

### PR DESCRIPTION
**Pull request by agedd**
### Summary of changes

- This PR merges the the Shopper Insights and MPE flow feature branch (`payment-insights-feature`)  into `main`. 

### Checklist

- [x] Added a changelog entry

- @agedd @scannillo

### CODE REVIEW

File: Demo/Application/Base/AppDelegate.swift [hunk 1]
```diff
@@ -28,6 +28,8 @@ import BraintreeCore
             userDefaults.set(BraintreeDemoAuthType.clientToken.rawValue, forKey: BraintreeDemoSettings.AuthorizationTypeDefaultsKey)
         } else if processInfoArgs.contains("-TokenizationKey") {
             userDefaults.set(BraintreeDemoAuthType.tokenizationKey.rawValue, forKey: BraintreeDemoSettings.AuthorizationTypeDefaultsKey)
+        } else if processInfoArgs.contains("-NewPayPalCheckoutTokenizationKey") {
```
Comment (scannillo) [inline Demo/Application/Base/AppDelegate.swift:31]: Can this be removed now? Not sure what it was even added for

File: Demo/Application/Features/PayPalWebCheckoutViewController.swift [hunk 2]
```diff
@@ -5,33 +5,78 @@ import BraintreePayPal
...
+        request.userAuthenticationEmail = emailTextField.text
+        
         let lineItem = BTPayPalLineItem(quantity: "1", unitAmount: "5.00", name: "item one 1234567", kind: .debit)
         lineItem.upcCode = "123456789"
         lineItem.upcType = .UPC_A
         lineItem.imageURL = URL(string: "https://www.example.com/example.jpg")
         request.lineItems = [lineItem]
+        
+        request.offerPayLater = payLaterToggle.isOn
+
```
Comment (jaxdesmarais) [inline Demo/Application/Features/PayPalWebCheckoutViewController.swift:78]:
```suggestion
```

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,132 @@
+import UIKit
+
+class BraintreeDemoPreferredPaymentMethodsViewController: BraintreeDemoBaseViewController {
+    private let preferredPaymentMethods: BTPreferredPaymentMethods
+    private let paypalClient: BTPayPalClient
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:5]:
```suggestion
    private let payPalClient: BTPayPalClient
```

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    private let venmoClient: BTVenmoClient
+    private let payPalCheckoutButton = UIButton(type: .system)
+    private let payPalVaultButton = UIButton(type: .system)
+    private let venmoButton = UIButton(type: .system)
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:9]: Can we use the `createButton` helper for all the buttons in here?

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import UIKit
+
+class BraintreeDemoPreferredPaymentMethodsViewController: BraintreeDemoBaseViewController {
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:3]: Ah, I see if we use `PaymentButtonBaseViewController` I think we will get access to a lot of helpers here for setting up buttons/the view/etc

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        view.addSubview(venmoButton)
+        
+        view.addConstraints([preferredPaymentMethodsButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+                             preferredPaymentMethodsButton.centerYAnchor.constraint(equalTo: view.centerYAnchor),
+                             payPalVaultButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+                             payPalVaultButton.bottomAnchor.constraint(equalTo: payPalCheckoutButton.bottomAnchor, constant: -40),
+                             payPalCheckoutButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+                             payPalCheckoutButton.bottomAnchor.constraint(equalTo: venmoButton.bottomAnchor, constant: -40),
+                             venmoButton.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+                             venmoButton.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -40)])
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:54]: It looks like the spacing here is off

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        button.isEnabled = false
+        
+        let paypalRequest = BTPayPalCheckoutRequest(amount: "4.30")
+        paypalClient.tokenizePayPalAccount(with: paypalRequest) { (nonce, error) in
+            button.isEnabled = true
+
+            if let e = error {
+                self.progressBlock(e.localizedDescription)
+            } else if let n = nonce {
+                self.nonceStringCompletionBlock(n.nonce)
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:84]:
```suggestion
            if let error {
                self.progressBlock(e.localizedDescription)
            } else if let nonce {
                self.nonceStringCompletionBlock(n.nonce)
```

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        
+        let paypalRequest = BTPayPalVaultRequest()
+        paypalRequest.activeWindow = self.view.window
+        paypalClient.tokenizePayPalAccount(with: paypalRequest) { (nonce, error) in
+            button.isEnabled = true
+            
+            if let e = error {
+                self.progressBlock(e.localizedDescription)
+            } else if let n = nonce {
+                self.nonceStringCompletionBlock(n.nonce)
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:105]:
```suggestion
            if let error {
                self.progressBlock(e.localizedDescription)
            } else if let nonce {
                self.nonceStringCompletionBlock(n.nonce)
```

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        button.isEnabled = false
+
+        let venmoRequest = BTVenmoRequest(paymentMethodUsage: .multiUse)
+        venmoClient.tokenizeVenmoAccount(with: venmoRequest) { (nonce, error) in
+            button.isEnabled = true
+            
+            if let e = error {
+                self.progressBlock(e.localizedDescription)
+            } else if let n = nonce {
+                self.nonceStringCompletionBlock(n.nonce)
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:125]:
```suggestion
            if let error {
                self.progressBlock(e.localizedDescription)
            } else if let nonce {
                self.nonceStringCompletionBlock(n.nonce)
```

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    }
+    
+    @objc func venmoButtonTapped(_ button: UIButton) {
+        self.progressBlock("Tapped Venmo")
+        
+        button.setTitle("Processing...", for: .disabled)
+        button.isEnabled = false
+
+        let venmoRequest = BTVenmoRequest(paymentMethodUsage: .multiUse)
+        venmoClient.tokenizeVenmoAccount(with: venmoRequest) { (nonce, error) in
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:119]:
```suggestion
        venmoClient.tokenizeVenmoAccount(with: venmoRequest) { nonce, error in
```

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    
+    @objc func payPalVaultButtonTapped(_ button: UIButton) {
+        self.progressBlock("Tapped PayPal Vault")
+        
+        button.setTitle("Processing...", for: .disabled)
+        button.isEnabled = false
+        
+        let paypalRequest = BTPayPalVaultRequest()
+        paypalRequest.activeWindow = self.view.window
+        paypalClient.tokenizePayPalAccount(with: paypalRequest) { (nonce, error) in
```
Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:99]:
```suggestion
        paypalClient.tokenizePayPalAccount(with: paypalRequest) { nonce, error in
```

Comment (jaxdesmarais) [inline Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:99]: Actually are these methods correct? They should all be `tokenize(request)`? This all looks like the old v5 syntax.

File: Demo/Application/Features/ShopperInsightsViewController.swift [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+        self.progressBlock("Tapped PayPal Vault")
+        shopperInsightsClient.sendPayPalSelectedEvent()
+        
+        button.setTitle("Processing...", for: .disabled)
+        button.isEnabled = false
+        
+        let paypalRequest = BTPayPalVaultRequest()
+        paypalRequest.userAuthenticationEmail = emailView.textField.text ?? nil
+        
+        paypalClient.tokenize(paypalRequest) { (nonce, error) in
```
Comment (jaxdesmarais) [inline Demo/Application/Features/ShopperInsightsViewController.swift:134]:
```suggestion
        paypalClient.tokenize(paypalRequest) { nonce, error in
```

File: Demo/Application/Features/ShopperInsightsViewController.swift [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    
+    @objc func venmoButtonTapped(_ button: UIButton) {
+        self.progressBlock("Tapped Venmo")
+        shopperInsightsClient.sendVenmoSelectedEvent()
+        
+        button.setTitle("Processing...", for: .disabled)
+        button.isEnabled = false
+
+        let venmoRequest = BTVenmoRequest(paymentMethodUsage: .multiUse)
+        venmoClient.tokenize(venmoRequest) { (nonce, error) in
```
Comment (jaxdesmarais) [inline Demo/Application/Features/ShopperInsightsViewController.swift:148]:
```suggestion
        venmoClient.tokenize(venmoRequest) { nonce, error in
```

File: Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift (hunk 11 above)
Comment (jaxdesmarais) [reply Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:99]: Seems like maybe this entire ViewController could be old and committed on accident?

Comment (scannillo) [reply Demo/Application/Features/Preferred Payment Methods/BraintreeDemoPreferredPaymentMethodsViewController.swift:99]: Yeh this VC is real old

File: Demo/Application/Features/ShopperInsightsViewController.swift (hunk 12 above)
Comment (scannillo) [reply Demo/Application/Features/ShopperInsightsViewController.swift:134]: Speaking of - I added a tickets for future us to move the Demo app & Unit Tests to use the SDKs async-await methods vs completion ones

File: Demo/Application/Base/AppDelegate.swift (hunk 1 above)
Comment (jaxdesmarais) [reply Demo/Application/Base/AppDelegate.swift:31]: It sounds like this feature is not enabled for our sandbox tokenization key on iOS. I had asked the same on the Android PR as well.

**Comment by jaxdesmarais** [comment]
Opened a PR with my suggestions here: https://github.com/braintree/braintree_ios/pull/1332

**Review by jaxdesmarais — APPROVED** [review]

**Review by tdchow — APPROVED** [review]

**Pull request merged**